feat: add inheritance-based validation and schemas

EdmilsonRodrigues opened the pull request:
- [X] Have you followed the [guidelines for contributing](https://github.com/canonical/snapcraft/blob/main/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [X] Have you successfully run `make test`?

---
I created the new inheritance based Project validation, with the simplified scope discussed in the clinics.
I took the charmcraft project as inspiration for the schema generation, and validation, used its pipeline as my base for the schema.yaml pipeline to assure that the schema file will always match the current state of the code.

In the charmcraft project, there is a validation of the schema, made using files from the smoketests directory.
We do not have a smoketests directory in the snapcraft project. Is there a place where there are files ready for this, should I add these files in a specific directory, or is it out of scope? For now that job is blocked so it won't be run.

In the `Makefile`, I only added the task to generate the schema. It is to `new_snapcraft.json` so the reviewer can check if the generated json is in the right format, having the old one as a comparison. If the schema generation is approved, I will make a commit changing to `snapcraft.json` as it should be.

In `snapcraft.models.__init__.py`, I only exported the new models, so they can be imported in the tests.

In `snapcraft.models.project.py` I created the project models, and the validation logic, and moved some validation from the monolithic model to the new models. Also, I overrode the unmarshal function to validate the data through the TypeAdapter of `SnapcraftProject`.

I had to use Discriminator functions, instead of field discriminators, because as this was an initial, with only three supported bases and two build bases, there should be a default fallback for the ones that are neither of them, but I used the enums, to generate them based on the Enum, as discussed.

In `conftest.py` I made a small change in the fixture `setup_project` to use unmarshall instead of model_validate, so the correct submodel will be used, as specific validations are made by them.

In `tests/unit/models/test_projects.py` I added a test for the unmarshaller, inside of the `TestProjectValidation` class, and changed a line that checked for "core22" base to check an instance of Core22Project, so mypy would not complain.

In `tests/unit/services/test_buildplan.py` I just changed a Project.model_validate to Project.unmarshall to use the correct submodel.

In `tools/schema.py` I based myself in the `tools/schema.py` of the charmcraft project.

This PR is still not ready to be merged, since I need some guidance in relation to the validate-schema job, and also if the generated schema is compliant with the expectations.

I would also like to know if there is anything missing, or anything that I got wrong so I can fix it. I did not change the docs, to document my change, because I was not sure if it is in the scope of this issue.

I am also sorry for the delay in this delivery, but I will be alert, waiting for the comments, and review. 


Fixes #5543

EdmilsonRodrigues commented:
I am terribly sorry, I commited a mistake here with the rebase. But I'll fix it right now.

mr-cal commented:
> I am also sorry for the delay in this delivery, but I will be alert, waiting for the comments, and review.

No need to apologize, this is a big effort and we don't have hard expectations on timing, as long as it doesn't go stale for weeks on end. We'll have to keep re-approving your workflow as you're a first-time contributor. Feel free to ping here and someone will approve the workflows.

> I am terribly sorry, I commited a mistake here with the rebase

No problem again. I ask that after this gets its first review, to not force-push anymore. That will make it easier to view changes between reviews.

mr-cal commented:
Fixes https://github.com/canonical/snapcraft/issues/5543

EdmilsonRodrigues commented:
> No problem again. I ask that after this gets its first review, to not force-push anymore. That will make it easier to view changes between reviews.

Alright!

mr-cal commented on snapcraft/models/project.py:
```diff
@@ -2199,6 +2156,123 @@ def get_partitions(self) -> list[str] | None:
+
+
+class Core22Project(BaseProject):
+    platforms: dict[str, Platform] | None = pydantic.Field(  # type: ignore[assignment,reportIncompatibleVariableOverride]
+        default=None,
+        description="The platforms key is only used in core24 and newer snaps. For core22 and older snaps, use "
+        "the ``architectures`` key.",
+    )
```
This shouldn't be allowed. This is causing:

```yaml
name: foo
base: core22
confinement: strict
platforms:
 amd64:
   build-on: [amd64]
   build-for: [amd64]

parts:
  my-part:
    plugin: nil
```

to be considered valid yaml. It shouldn't be, because `platforms` and `core22` are mutually exclusive.

EdmilsonRodrigues replied:
I understand! But removing this causes 17 tests in `test/unit/commands/test_remote.py` to fail for the reason: `AssertionError: Expected 'start_builds' to be called once. Called 0 times.`

I checked it last week and it happens because there is an error unmarshalling the Project in these tests, because their data has platforms, and so they cannot be parsed as Core22 projects. I will check them again in order to bring better data about it. I might be sending the updated info tomorrow morning.

EdmilsonRodrigues replied:
I've got these changes stashed, so as soon as there's a way to run these tests, I'll be ready to commit it!

lengau replied:
The project model does need to accept a `platforms` key, but we don't have to make that fact a part of the schema. Something like this for the core22 project should do the trick:

```python
from pydantic.json_schema import SkipJsonSchema

platforms: SkipJsonSchema[dict[str, Platform] | None] = pydantic.Field(
    default=None,
    description="Not available for core22. Use the `architectures` key.",
	exclude=True,
	repr=False,
)
```

This will make it unavailable in the JSON schema (so the example YAML would be considered invalid under the schema). Compare to how the [BasesCharm](https://github.com/canonical/charmcraft/blob/2ddafd0e9d61c99a47740fcb5ecbad602c2f08ff/charmcraft/models/project.py#L810C1-L816C6) model works.

Then in the `ProjectService`, we can have `_app_preprocess_project` we can check that base `core22` and build-base `core22` are both incompatible with the existence of a `platforms` key.

It's a little bit of a bodge, but it only applies to "legacy" snaps anyway.

mr-cal commented on snapcraft/models/project.py:
```diff
@@ -2199,6 +2156,123 @@ def get_partitions(self) -> list[str] | None:
+        return validate_architectures(architectures)
+
+
+class BareCore22Project(Core22Project): ...
```
This should require `base: bare`.

EdmilsonRodrigues replied:
Alright! I am adding it right away!

mr-cal commented on snapcraft/models/project.py:
```diff
@@ -2199,6 +2156,123 @@ def get_partitions(self) -> list[str] | None:
+    )
+
+
+class BareCore24Project(Core24Project): ...
```
Also should require `base: bare`.

mr-cal commented on snapcraft/models/project.py:
```diff
@@ -2199,6 +2156,123 @@ def get_partitions(self) -> list[str] | None:
+    return pydantic.WrapValidator(_validator)
+
+
+class BaseProject(Project): ...
```
I don't think we need this?

EdmilsonRodrigues replied:
Perfect! Thank you! Fixing it right away!

mr-cal commented on snapcraft/models/project.py:
```diff
@@ -2199,6 +2156,123 @@ def get_partitions(self) -> list[str] | None:
+class BaseProject(Project): ...
+
+
+class BareProject(BaseProject): ...
```
I'm not sure about this. A bare project should be defined by the bare core22 and bare core24 classes.

mr-cal reviewed (commented):
Approach looks good so far, but the hierarchy needs some work.

The schema isn't matching correctly. For example,

```yaml
name: foo
base: core24
confinement: strict
platforms:
 amd64:
   build-on: [amd64]
   build-for: [amd64]

parts:
  my-part:
    plugin: nil
```

appears to be matching multiple anyOf in the new schema and raising a `Matches multiple schemas when only one must validate` yaml schema error.

EdmilsonRodrigues commented on .github/workflows/schema.yaml:
```diff
@@ -0,0 +1,66 @@
+
+      - name: Run generator script
+        run: |
+          uv run --frozen \
+            python tools/schema.py > generated_schema.json
```
Do you think it would be better?

EdmilsonRodrigues replied:
Alright! Thank you for your answer!

EdmilsonRodrigues commented on .github/workflows/schema.yaml:
```diff
@@ -0,0 +1,66 @@
+    if: false
+    runs-on: ubuntu-latest
+    env:
+      SPREAD_FILES_GLOB: tests/spread/smoketests/basic/charmcraft* #! Must be replaced
```
So should I sign off and commit this suggestion also?

EdmilsonRodrigues commented:
> appears to be matching multiple anyOf in the new schema and raising a `Matches multiple schemas when only one must validate` yaml schema error.

As today's my birthday, I'll come back to this problem, as well as to the schema validation that is failing, tomorrow. If there is an any more comment or answer, I'll also review it tomorrow, and commit the solution as soon as possible!!

Thank you all for the responses, and for guiding me through this PR!!

mr-cal commented on new_schema/snapcraft.json:
Will you move this to `schema/snapcraft.json`? Now that #5744 is completed, we can overwrite this existing file.

mr-cal reviewed (commented):
> As today's my birthday

Hi, happy belated birthday!

This is looking good, but I'm still getting a `Matches multiple schemas when only one must validate.` error in my editor.

EdmilsonRodrigues commented:
> Hi, happy belated birthday!

Thank you!!

mr-cal commented:
> There was a confusion in my part, as lint is not one of the conventional commit tag. I amended it in my local, but to overwrite I would need to force push. Should I do it? Sorry about this one.

No problem. It's an optional CI check and this PR will be a squash-commit anyway.

EdmilsonRodrigues commented:
> No problem. It's an optional CI check and this PR will be a squash-commit when it's merged.
Alright, thank you! I see that the ajv run is still not running. I will install it and debug it here, maybe this is the case of this error:

> This is looking good, but I'm still getting a Matches multiple schemas when only one must validate. error in my editor.

bepri commented:
> But I do not have an extension that allows me to validate yaml based on an schema, I am also curious about how to do it.

There's two ways you could go about this - the manual way (quick setup, but annoying to use) or the automated way (maybe difficult setup, but instant from then on out).

The manual way would just be to use a CLI tool like [jsonschema](https://github.com/sourcemeta/jsonschema) (which has a snap!):
```
$ jsonschema validate schema/schema.json snapcraft.yaml
```

Or, you could configure Emacs to support LSP and then use a YAML language server. It seems Emacs LSP has [a guide for this](https://emacs-lsp.github.io/lsp-mode/page/lsp-yaml/). However, as a fair disclaimer, I've never used Emacs LSP.

EdmilsonRodrigues commented:
> Or, you could configure Emacs to support LSP and then use a YAML language server. It seems Emacs LSP has [a guide for this](https://emacs-lsp.github.io/lsp-mode/page/lsp-yaml/). However, as a fair disclaimer, I've never used Emacs LSP.

Thank you for the help, I'll configure it later, and thank you also by the tool recommendation!

I'll check it here in order to validate it! And as soon as it is done I'll commit it! 

If I do not commit the fix tomorrow, I'll do it until Monday morning!!

EdmilsonRodrigues commented:
@mr-cal Sorry again for the time I took, it was a little harder than I initially thought.

About the schema validation, I fixed all validation errors, but one:
App does not have an extension parameter, so it fails for additional properties not allowed. 
I do not know how to proceed with this one.

This makes an error in the schemas:
- tests/spread/core22/linters/library/snap/snapcraft.yaml
- tests/spread/core22/linters/linter-ros2-humble-mixed/snap/snapcraft.yaml
- tests/spread/core22/linters/linter-ros2-humble/snap/snapcraft.yaml
- tests/spread/core24/linters-pack/library-gnome/snap/snapcraft.yaml

I had to do a few modifications to some schemas. Specifically:
 Core24Project:
 Changes:
- dict[str, Platform | None] | None 
What it enabled to work:
```yaml
platforms:
  platform1:
```

Platform: 
Changes:
- build_on: UniqueList[str] | str | None
- build_for: SingleEntryList | str | None
What it enabled to work
```yaml
platforms:
  platform1:
    build-on:
    build-for:
  platform2:
    build-on: "amd64"
    build-for: "amd64"
```

Project:
Changes:
- website: UniqueList[str] | str | None
- issues: UniqueList[str] | str | None
- source_code: UniqueList[str] | str | None
- donation: UniqueList[str] | str | None
It enabled to work the options with only one instead of an array.

I also changed how the validation worked, from the discriminator functions I used before, to one dependent on base, and removed the bigger Project from the schemas generated, because this was a catch-all, and was the reason for the others matching two schemas.

Also, I created a BaseCore22Project and BaseCore24Project, for the snaps with type "base". It was needed because there were some schemas in the spread of type base, and they were not being correctly matched, since base is a required field for the others.

bepri commented:
Thanks! You can ignore the failing spread test related to flutter, that one isn't related to this change. I'd also take a look at the tasks failing under the "validate-schema" workflow again, those seem relevant.

> Sorry again for the time I took

No need to apologize :) you're actively working on this task, and I think that's good enough!

EdmilsonRodrigues commented:
> No need to apologize :) you're actively working on this task, and I think that's good enough!

Thank you!



> I'd also take a look at the tasks failing under the "validate-schema" workflow again, those seem relevant.

I need a little guidance in this one.

It is related to the App model doesn't have an "extensions" property. So they are failing because no extra property is allowed.

Should I add this property to the App model?

bepri commented:
Interesting that it was missing. Yes, please do add it! For validation, [this function](https://github.com/canonical/snapcraft/blob/main/snapcraft/extensions/registry.py#L59) should always return the list of valid extensions.

EdmilsonRodrigues commented:
> Interesting that it was missing. Yes, please do add it! For validation, [this function](https://github.com/canonical/snapcraft/blob/main/snapcraft/extensions/registry.py#L59) should always return the list of valid extensions.

Alright! Thank you!

I'll commit it ASAP! It may take a while, because today I'm a little busy, but I think I'll still commit the fixes today!

EdmilsonRodrigues commented:
@mr-cal @brepi @lengau 

So, what do you think? 
Do these changes are a good fix for the issue? 

If there's anything more to add or change, I'll be glad to do it, in order to deliver the best patch possible in this contrib!! It's my first, but definitely not last contrib to snapcraft, and I want it to be the best possible!

bepri commented on snapcraft/parts/update_metadata.py:
```diff
@@ -257,5 +257,5 @@ def _raise_formatted_validation_error(err: pydantic.ValidationError):
         return
 
-    varname = ".".join(x for x in loc if isinstance(x, str))
+    varname = ".".join(x for x in loc[2:] if isinstance(x, str))
```
This makes sense. Could you add a comment in the code explaining this, with the example of `grade` vs `UNIMPLEMENTED.core22.grade`?

bepri commented on snapcraft/models/project.py:
```diff
@@ -1019,6 +1043,17 @@ def _validate_aliases(cls, aliases: list[str]) -> list[str]:
+            if extension not in valid_extensions:
+                raise ValueError(
+                    f"{extension!r} is not a valid extension. Valid extensions are {valid_extensions!r}."
+                )
```
```suggestion
        invalid_extensions = [ext for ext in extensions if ext not in valid_extensions]
        if invalid_extensions:
            raise ValueError(
                f"The following extensions are invalid: {invalid_extensions!r}.\n"
                f"Valid extensions are {valid_extensions!r}."
            )
```

bepri commented on snapcraft/models/project.py:
```diff
@@ -2218,6 +2209,205 @@ def get_partitions(self) -> list[str] | None:
+        ]
+    ] = pydantic.Field(
+        default=None,
+        description="The architectures key is only used in core22 snaps. For core24 and newer snaps, use the ``platforms`` key.",
```
```suggestion
        description="The architectures key is only used in core22 snaps and below. For core24 and newer snaps, use the ``platforms`` key.",
```

I know this PR is ignoring core20 and lower, but this is might've caused some confusion in the future.

EdmilsonRodrigues commented on .github/workflows/schema.yaml:
```diff
@@ -0,0 +1,64 @@
+  validate-schema:
+    runs-on: ubuntu-latest
+    env:
+      SPREAD_FILES_GLOB: tests/spread/core2{2,4}/**/snapcraft.yaml
```
Then, should we also add a `Core26Project`, `BareCore26Project` and `BaseCore26Project` to our models?

EdmilsonRodrigues commented on snapcraft/models/project.py:
```diff
@@ -40,12 +41,14 @@
 )
 from craft_platforms import DebianArchitecture
 from pydantic import ConfigDict, PrivateAttr, StringConstraints, error_wrappers
+from pydantic.json_schema import SkipJsonSchema  # noqa: TC002
```
I'll sure try it! 
But I think this lint was the one that alerts when the import is only used for typing, and the lint asks to put inside of the `if TYPE_CHECKING`, but if I do this, the Json schema will not have them skipped.

But I'll look into it to be sure, and think about a way of fixing it without needing to silence it.

bepri replied:
Interesting... if that's the case, even after checking with the `from __future___ ...` import, then go ahead and keep the `noqa`. I think that would be a fine justification. Thanks for checking!

EdmilsonRodrigues commented on snapcraft/models/project.py:
```diff
@@ -2218,6 +2209,190 @@ def get_partitions(self) -> list[str] | None:
+
+class Core22Project(Project):
+    type: Annotated[  # type: ignore[assignment,reportIncompatibleVariableOverride]
+        Literal["app", "gadget", "kernel", "snapd", None],
```
Sure!

I have these changes stashed now!
All tests worked perfectly, but some of lifecycle!

That is because of this function in craft-platforms:
https://github.com/canonical/craft-platforms/blob/main/craft_platforms/snap/_build.py#L92-L163

This function requires snap_type as a `str`, but now they are `ProjectType`. (Name of the enum)

I am working on this to pass snap_type as a str right now, that's the only reason I didn't pushed these changes yet. But I plan to send them until the end of the day!

EdmilsonRodrigues commented:
About the QA test that is failing, it is this test:
```python
FAILED tests/unit/parts/test_setup_assets.py::TestSetupAssets::test_setup_assets_remote_icon - requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
```
But all tests from this file were tested in passed in the fast-tests. 
I do not know exactly why this problem hapenned. But I'll investigate it through the day

lengau reviewed (approved):
Looks great, thank you!

EdmilsonRodrigues commented:
Thank you all, you are all so kind!! :)
